### PR TITLE
feat: some tikv todo impl

### PR DESCRIPTION
**Pull request by zyxxoo**

### CODE REVIEW

**Review by zhoney — APPROVED** [review]

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java [hunk 1]
```diff
@@ -19,10 +19,31 @@
...
+import org.tikv.common.TiSession;
+import org.tikv.common.key.Key;
+import org.tikv.common.region.TiRegion;
+import org.tikv.common.util.BackOffer;
+import org.tikv.common.util.ConcreteBackOffer;
+import org.tikv.common.util.FastByteComparisons;
+import org.tikv.kvproto.Kvrpcpb;
+import org.tikv.raw.RawKVClient;
+import org.tikv.shade.com.google.protobuf.ByteString;
+
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java]: can keep the origin import order?

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java [hunk 2]
```diff
@@ -280,32 +276,33 @@ public TikvShardSpliter(String table) {
             super(table);
         }
 
+        private String startKey(byte[] start) {
+            return Arrays.equals(start, START_BYTES) ?
+                    START : StringEncoding.encodeBase64(start);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java]: prefer to align with Arrays

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java [hunk 3]
```diff
@@ -254,9 +266,72 @@ public void rollback() {
...
+                                 return Pair.of(this.originKey(prefix,
+                                                     region.getStartKey()
+                                                           .toByteArray()),
+                                                this.originKey(prefix,
+                                                               realEndKey));
+                           }).collect(Collectors.toList());
+        }
+
+        public byte[] originKey(byte[] prefix, byte[] regionKey) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java]: pass as `int tablePrefixLength` param?

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java [hunk 4]
```diff
@@ -254,9 +266,72 @@ public void rollback() {
...
+                                     realEndKey = region.getEndKey()
+                                                        .toByteArray();
+                                 }
+
+                                 return Pair.of(this.originKey(prefix,
+                                                     region.getStartKey()
+                                                           .toByteArray()),
+                                                this.originKey(prefix,
+                                                               realEndKey));
+                           }).collect(Collectors.toList());
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java:327]: can improve the code for more readable?

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java (hunk 3 above)
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java]: can mark as private static as possible

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java [hunk 5]
```diff
@@ -347,8 +423,8 @@ private ByteString key(String table) {
 
         private ByteString key(String table, byte[] key) {
             byte[] prefix = table.getBytes();
-            byte[] actualKey = new byte[prefix.length + 1 + key.length];
-            System.arraycopy(prefix, 0, actualKey, 0, prefix.length);
+            byte[] actualKey = Arrays.copyOf(prefix,
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java:444]: could you update the alignment of scan() method, seems exceed 80 chars

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java [hunk 6]
```diff
@@ -168,7 +169,7 @@ protected BackendColumnIterator queryBy(Session session, Query query) {
             assert !query.ids().isEmpty();
             // NOTE: this will lead to lazy create iterator
             return new BackendColumnIteratorWrapper(new FlatMapperIterator<>(
-                    query.ids().iterator(), id -> this.queryById(session, id)
+                    query.ids().iterator(), id -> this.getById(session, id)
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java]: not sure whether the core-test can pass after this switch

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java (hunk 4 above)
Comment (zyxxoo) [reply src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java:327]: I have added some annotation

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java (hunk 6 above)
Comment (zyxxoo) [reply src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvTable.java]: ok, I have restored

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java [hunk 7]
```diff
@@ -157,6 +170,19 @@ public static final String decode(byte[] bytes) {
...
+    private static byte[] originKey(int tablePrefixLength,
+                                    byte[] regionKey) {
+        if (regionKey == null || regionKey.length == 0) {
+            return new byte[0];
+        }
+
+        assert tablePrefixLength < regionKey.length;
+        byte[] originKey = new byte[regionKey.length - tablePrefixLength];
+        System.arraycopy(regionKey, tablePrefixLength,
+               originKey, 0, originKey.length);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java]: can align with regionKey?

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java [hunk 8]
```diff
@@ -254,9 +280,73 @@ public void rollback() {
...
-            return null;
+        public List<Pair<byte[], byte[]>> keyRanges(String table) {
+            E.checkArgument(StringUtils.isNotBlank(table),
+                            "Table don't allow empty");
+
+            ByteString startKey = this.key(table);
+            ByteString endKey = ByteString.copyFrom(
+                                Key.toRawKey(startKey).nextPrefix().getBytes());
+            List<TiRegion> regions = this.fetchRegionsFromRange(startKey,
+                                                                endKey);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java:291]: may fetchRegionsFromRange return a region which is out of the endKey?

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java [hunk 9]
```diff
@@ -254,9 +280,73 @@ public void rollback() {
...
+                                *  ...
+                                * regionN [startKeyN, regionEndKeyN]
+                                * if regionEndKeyN < tableEndKey, the table real
+                                * end key is regionEndKeyN, otherwise the real
+                                * end key is table end key
+                                */
+                                 byte[] realEndKey = null;
+                                 if (FastByteComparisons.compareTo(
+                                     endKey.toByteArray(),
+                                     region.getEndKey().toByteArray()) < 0) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java]: let region.getEndKey() first when compareTo()?
and can use `region.getEndKey().compareTo(endKey) > 0`

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java (hunk 8 above)
Comment (zyxxoo) [reply src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java:291]: if return an out of the endkey, the query will return other table data when use the out of endkey params

File: src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java (hunk 9 above)
Comment (zyxxoo) [reply src/main/java/com/baidu/hugegraph/backend/store/tikv/TikvStdSessions.java]: The region type is TiRegion,  and the type is no the compareTo method

**Review by javeme — APPROVED** [review]

**Pull request merged**